site/content/extend/plugins/signing-plugins.md

```diff
@@ -0,0 +1,151 @@
+```
+
+
+### Importing the Key
+If you already have a public and private key pair, you can import them to the GPG.
+```console
+> gpg --import ./my-priv-gpg-key
+> gpg --import ./my-pub-gpg-key
+```
+
+
+## Plugin Signing
+For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace `XXXXXXXX` with your hexadecimal key ID and run:
+```console
+> gpg -u XXXXXXXX --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
+```
+This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig`, which is the signature of your plugin.
+
+## Plugin Verification
+Mattermost server will verify plugin signatures downloaded from Mattermost Marketplace. Mattermost public key is shipped with the server. If needed one can download and verify plugins from custom marketplace as well. To do this, custom public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:
+```console
+> mattermost plugin add key my-pub-key
+```
+
+Multiple public keys can be added to the Mattermost server:
+```console
+> mattermost plugin add key my-pk-file1 my-pk-file2
+```
+To list the names of all public keys installed on your Mattermost server, use:
+```console
+> mattermost plugin keys
+```
+To delete public key(s) from your Mattermost server, use:
+```console
+> mattermost plugin delete key my-pk-file1 my-pk-file2
+```
+
+### Storing Plugin Signatures
+When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore (not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures alongside plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format. Which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.
+```console
+# Filestore plugin structure
+/plugins/<plugin_id>.tar.gz            # plugin bundle tar
+/plugins/<plugin_id>.<counter-1>.sig   # multiple plugin signatures
+/plugins/<plugin_id>.<counter-2>.sig     
+
+# Examples
+/plugins/com.mattermost.demo.tar.gz    # plugin bundle tar
+/plugins/com.mattermost.demo.1.sig     # multiple plugin signatures
+/plugins/com.mattermost.demo.2.sig
+/plugins/com.mattermost.demo.3.sig
+/plugins/com.github.matterpoll.tar.gz
+/plugins/com.github.matterpoll.1.sig
+/plugins/com.github.matterpoll.2.sig
+/plugins/com.github.matterpoll.3.sig
+
+```
+
+### Mattermost Plugin Assets Naming Convention
```

Comment:
This seems unrelated to the scope of this document. Perhaps this might be better located in the marketplace documentation itself for how we expect the generator to work.